File: frontend/WhereIsMyMoneyUI/src/app/models/rule/Rule.ts
export type MatchType = 'Exact' | 'Partial' | 'Regex';

export interface Rule {
  id: number;
  accountId: number;
  name: string;
  priority: number;
  isActive: boolean;
  matchType: MatchType;
  descriptionPattern: string;
  minAmount?: number;
  maxAmount?: number;
  budgetId?: number;
  daysOfWeek?: number[];
  dayOfMonth?: number;
  categoryIds: number[];
}

export interface CreateRuleRequest {
  name: string;
  matchType: MatchType;
  descriptionPattern: string;
  minAmount?: number;
  maxAmount?: number;
  budgetId?: number;
  daysOfWeek?: number[];
  dayOfMonth?: number;
  categoryIds: number[];
}

export interface UpdateRuleRequest extends CreateRuleRequest {
  isActive: boolean;
  priority: number;
}

export interface PatchRuleRequest {
  isActive?: boolean;
  priority?: number;
  name?: string;
  matchType?: MatchType;
  descriptionPattern?: string;
  minAmount?: number;
  maxAmount?: number;
  budgetId?: number;
  daysOfWeek?: number[];
  dayOfMonth?: number;
  categoryIds?: number[];
}

export interface ApplyToExistingRequest {
  fromDate: string;
  toDate: string;
  overwriteExisting: boolean;
}

File: backend/src/WhereIsMyMoney.Api/Models/RuleModels/Rule.cs
namespace WhereIsMyMoney.Api.Models.RuleModels;

public sealed class Rule
{
    public long Id { get; set; }
    public long AccountId { get; set; }
    public required string Name { get; set; }
    public int Priority { get; set; }
    public bool IsActive { get; set; } = true;
    public MatchType MatchType { get; set; }
    public required string DescriptionPattern { get; set; }
    public decimal? MinAmount { get; set; }
    public decimal? MaxAmount { get; set; }
    public long? BudgetId { get; set; }
    public int[]? DaysOfWeek { get; set; }
    public int? DayOfMonth { get; set; }
    public int[] CategoryIds { get; set; } = [];
}

public enum MatchType
{
    Exact,
    Partial,
    Regex,
}
